repos/system_upgrade/el7toel8/actors/migratentp/libraries/ntp2chrony.py

```diff
@@ -0,0 +1,670 @@
+
+        conf += "# Record the rate at which the system clock gains/losses time.\n"
+        if not self.driftfile:
+            conf += "#"
+        conf += "driftfile /var/lib/chrony/drift\n"
+        conf += "\n"
+
+        conf += "# Allow the system clock to be stepped in the first three updates\n"
+        conf += "# if its offset is larger than 1 second.\n"
+        conf += "makestep 1.0 3\n"
+        conf += "\n"
+
+        conf += "# Enable kernel synchronization of the real-time clock (RTC).\n"
+        conf += "rtcsync\n"
+        conf += "\n"
+
+        conf += "# Enable hardware timestamping on all interfaces that support it.\n"
+        conf += "#hwtimestamp *\n"
+        conf += "\n"
+
+        if maxdistance > 0.0:
+            conf += "# Specify the maximum distance of sources to be selectable.\n"
+            conf += "maxdistance {}\n".format(maxdistance)
+            conf += "\n"
+
+        conf += "# Increase the minimum number of selectable sources required to adjust\n"
+        conf += "# the system clock.\n"
+        if minsources > 1:
+            conf += "minsources {}\n".format(minsources)
+        else:
+            conf += "#minsources 2\n"
+        conf += "\n"
+
+        conf += self.get_chrony_conf_allows()
+
+        conf += self.get_chrony_conf_cmdallows()
+
+        conf += "# Serve time even if not synchronized to a time source.\n"
+        if orphan_stratum > 0 and orphan_stratum < 16:
+            conf += "local stratum {} orphan\n".format(orphan_stratum)
+        elif local_stratum > 0 and local_stratum < 16:
+            conf += "local stratum {}\n".format(local_stratum)
+        else:
+            conf += "#local stratum 10\n"
+        conf += "\n"
+
+        conf += "# Specify file containing keys for NTP authentication.\n"
+        conf += "keyfile {}\n".format(chrony_keys_path)
+        conf += "\n"
+
+        conf += "# Get TAI-UTC offset and leap seconds from the system tz database.\n"
+        conf += "leapsectz right/UTC\n"
+        conf += "\n"
+
+        conf += "# Specify directory for log files.\n"
+        conf += "logdir /var/log/chrony\n"
+        conf += "\n"
+
+        conf += "# Select which information is logged.\n"
+        if logs:
+            conf += "log {}\n".format(" ".join(logs))
+        else:
+            conf += "#log measurements statistics tracking\n"
+
+        return conf
+
+    def get_chrony_keys(self):
+        if not self.keyfile:
+            return ""
+
+        keys = "# This file was converted from {}.\n".format(self.keyfile)
+        keys += "\n"
+
+        for key in self.keys:
+            id = key[0]
+            type = key[1]
+            password = key[2]
+
+            if type in ["m", "M"]:
+                type = "MD5"
+            elif type not in ["MD5", "SHA1", "SHA256", "SHA384", "SHA512"]:
+                continue
+
+            prefix = "ASCII" if len(password) <= 20 else "HEX"
+
+            for first, last in self.trusted_keys:
```

Comment:
iiuc L600-L605 are the same as
trusted = any(id in range(first, last+1) for (first, last) in self.trusted_keys)

Reply:
Similar, but faster for wide ranges and maybe slightly more readable.